ludwig/utils/tracker.py

```diff
@@ -0,0 +1,169 @@
+        try:
+            message = queue.get(block=False)
+            if isinstance(message, str):
+                if message == STOP_MESSAGE:
+                    save_json(os.path.join(output_dir, info["tag"] + "_temp.json"), info)
+                    return
+            else:
+                queue.put(message)
+        except EmptyQueueException:
+            pass
+
+        gpu_infos = GPUStatCollection.new_query()
+        for i, gpu_info in enumerate(gpu_infos):
+            gpu_key = f"gpu_{i}"
+            info["system"][gpu_key]["memory_used"].append(gpu_info.memory_used)
+        info["system"]["cpu_utilization"].append(psutil.cpu_percent())
+        info["system"]["ram_utilization"].append(psutil.virtual_memory().percent)
+        info["system"]["swap_utilization"].append(psutil.swap_memory().percent)
+        time.sleep(logging_interval)
+
+
+class Tracker:
+    """Track system resource (hardware and software) usage by a chunk of code."""
+
+    def __init__(self, tag, output_dir, logging_interval=1, num_batches=None, num_examples=None):
+        """
+        tag: one of `train` or `evaluate`.
+        output_dir: path where metrics are saved.
+        logging_interval: time interval in seconds at which system is polled for resource usage.
+        num_batches: number of batches of training or evaluation process.
+        num_batches: number of examples of training or evaluation process.
+        """
+        self.output_dir = output_dir
+        self.tag = tag
+        self.info = {"tag": self.tag, "system": {}}
+        self.num_batches = num_batches
+        self.num_examples = num_examples
+        self.logging_interval = logging_interval
+        self.launched = False
+        os.makedirs(os.path.join(self.output_dir), exist_ok=True)
+
+    def populate_static_information(self):
+        """Populates the report with static software and hardware information."""
+        self.info["ludwig_version"] = LUDWIG_VERSION
+        self.info["start_disk_usage"] = shutil.disk_usage(os.path.expanduser("~")).used
+
+        # CPU information
+        self.info["system"]["python_packages_and_versions"] = [
+            str(package) for package in get_python_packages_and_versions()
+        ]
+        cpu_info = get_my_cpu_info()
+        self.info["system"]["cpu_architecture"] = cpu_info["arch"]
+        self.info["system"]["num_cpu"] = cpu_info["count"]
+        self.info["system"]["cpu_name"] = cpu_info["brand_raw"]
+
+        # GPU information
+        gpu_infos = get_gpu_info()
+        for i, gpu_info in enumerate(gpu_infos):
+            gpu_key = f"gpu_{i}"
+            self.info["system"][gpu_key] = {}
+            self.info["system"][gpu_key]["name"] = gpu_info["name"]
+            self.info["system"][gpu_key]["total_memory"] = gpu_info["total_memory"]
+            self.info["system"][gpu_key]["driver_version"] = gpu_info["driver_version"]
+            self.info["system"][gpu_key]["cuda_version"] = gpu_info["cuda_version"]
```

Comment:
Is it possible to get GPU utilization here? That's a key metric I look for while benchmarking.

Reply:
added in `monitor()`
`info["system"][gpu_key]["memory_used"].append(gpu_info.memory_used)`